core/domain/exp_domain.py

```diff
@@ -474,6 +474,246 @@ def __init__(self, change_list):
+        if version not in self:
+            raise KeyError('Span does not cover version=%r' % version)
+        return self._version_snapshots[version]
+
+    def __setitem__(self, unused_version, unused_snapshot):
+        """Raises an exception. Only `extend_or_split` can mutate a span."""
+        raise Exception('Only `extend_or_split` can mutate a span')
+
+    def __len__(self):
+        """Returns the number of versions covered by self."""
+        return len(self._version_snapshots)
+
+    @property
+    def names(self):
+        """Yields all (version, state name) pairs that this span holds."""
+        for version, (name, unused_state) in self:
+            yield (version, name)
+
+    @property
+    def states(self):
+        """Yields all (version, state) pairs that this span holds."""
+        for version, (unused_name, state) in self:
+            yield (version, state)
+
+    def get_version(self, version):
+        """Returns the state as it appeared in the given exploration version.
+
+        Args:
+            version: int. The version to get.
+
+        Returns:
+            state_domain.State. The state as it appeared in the given version.
+        """
+        unused_name, state = self[version]
+        return state
+
+    def get_version_name(self, version):
+        """Returns the name of the state at the given exploration version.
+
+        Args:
+            version: int. The version to get.
+
+        Returns:
+            str. The name of the state at the given exploration version.
+        """
+        name, unused_state = self[version]
+        return name
+
+    def get_multi_versions(self, version_start, version_end):
+        """Returns states for a half-open range of exploration versions.
+
+        Args:
+            version_start: int. The lower bound of versions to get (inclusive).
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(state_domain.State). The states corresponding to the given
+            range of versions.
+        """
+        return [
+            state for unused_name, state in (
+                self._get_multi_versions(version_start, version_end))
+        ]
+
+    def get_multi_version_names(self, version_start, version_end):
+        """Returns state names for a half-open range of exploration versions.
+
+        Args:
+            version_start: int. The lower bound of versions to get (inclusive).
+            version_end: int. The upper bound of versions to get (exclusive).
+
+        Returns:
+            list(str). The names corresponding to the given range of versions.
+        """
+        return [
+            name for name, unused_state in (
+                self._get_multi_versions(version_start, version_end))
+        ]
+
+    def extend_or_split(self, exp_version, state_name, state, exp_version_diff):
+        """Extends span to include the given state's version if it is compatible
```

Comment:
I don't think the concept of compatibility should be part of this domain object, which seems to be more framed around describing the history of the state and not the things the state contains. Adding a notion of compatibility here mixes things up, and I don't really get the semantics of "split and return a new span" if the state is incompatible with the previous one.

Can the compatibility stuff be done outside this domain object?

Reply:
It already has been, see `state_domain.State.is_structurally_compatible_with`.

Reply:
No, not what I mean -- I think what you're calling "structurally compatible" here is a notion that lies more in the realm of stats rather than exploration structure. So that concept should be part of stats_services or whatever method ends up using this, not inbuilt into here.

Reply:
Got it, I've lifted it into a "split_predicate" function. I'd like to keep the splitting as an implementation detail of the span since it makes code more elegant. Call sites can provide None (to never split), or any callable predicate (examples can be found in the unit tests).

WDYT?